src/app/icd/client/DefaultICDClientStorage.cpp

```diff
@@ -0,0 +1,436 @@
+                                                         backingBuffer.Get(), length);
+    if (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND)
+    {
+        return CHIP_NO_ERROR;
+    }
+    ReturnErrorOnFailure(err);
+
+    TLV::ScopedBufferTLVReader reader(std::move(backingBuffer), len);
+
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::AnonymousTag()));
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(reader.EnterContainer(arrayType));
+
+    while ((err = reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag())) == CHIP_NO_ERROR)
+    {
+        ICDClientInfo clientInfo;
+        TLV::TLVType ICDClientInfoType;
+        NodeId nodeId;
+        FabricIndex fabric;
+        ReturnErrorOnFailure(reader.EnterContainer(ICDClientInfoType));
+        // Peer Node ID
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(ClientInfoTag::kPeerNodeId)));
+        ReturnErrorOnFailure(reader.Get(nodeId));
+
+        // Fabric Index
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(ClientInfoTag::kFabricIndex)));
+        ReturnErrorOnFailure(reader.Get(fabric));
+        clientInfo.peer_node = ScopedNodeId(nodeId, fabric);
+
+        // Start ICD Counter
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(ClientInfoTag::kStartICDCounter)));
+        ReturnErrorOnFailure(reader.Get(clientInfo.start_icd_counter));
+
+        // Offset
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(ClientInfoTag::kOffset)));
+        ReturnErrorOnFailure(reader.Get(clientInfo.offset));
+
+        // MonitoredSubject
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(ClientInfoTag::kMonitoredSubject)));
+        ReturnErrorOnFailure(reader.Get(clientInfo.monitored_subject));
+
+        // Shared key
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(ClientInfoTag::kSharedKey)));
+        ByteSpan buf(clientInfo.mSharedKey.AsMutable<Crypto::Aes128KeyByteArray>());
+        ReturnErrorOnFailure(reader.Get(buf));
+        memcpy(clientInfo.mSharedKey.AsMutable<Crypto::Aes128KeyByteArray>(), buf.data(), sizeof(Crypto::Aes128KeyByteArray));
+        ReturnErrorOnFailure(reader.ExitContainer(ICDClientInfoType));
+        clientInfoVector.push_back(clientInfo);
+    }
+
+    if (err != CHIP_END_OF_TLV)
+    {
+        return err;
+    }
+
+    return reader.ExitContainer(arrayType);
+}
+
+CHIP_ERROR DefaultICDClientStorage::SetKey(ICDClientInfo & clientInfo, const ByteSpan keyData)
+{
+    VerifyOrReturnError(keyData.size() == sizeof(Crypto::Aes128KeyByteArray), CHIP_ERROR_INVALID_ARGUMENT);
+
+    Crypto::Aes128KeyByteArray keyMaterial;
+    memcpy(keyMaterial, keyData.data(), sizeof(Crypto::Aes128KeyByteArray));
+
+    return mpKeyStore->CreateKey(keyMaterial, clientInfo.mSharedKey);
+}
+
+CHIP_ERROR DefaultICDClientStorage::Save(TLV::TLVWriter & writer, const std::vector<ICDClientInfo> & clientInfoVector)
```

Comment:
This should be called "SerializeToTlv" since it doesn't actually write the entry into storage.

Reply:
updated, thanks